### PR TITLE
Batch `MLFlowLogger` requests

### LINKED ISSUE #15900: Batch requests in `MLFlowLogger`
## Proposed refactor

Update the mlflow API callsites in `MLFlowLogger` to use the more efficient [`MlflowcClient.log_batch`](https://github.com/mlflow/mlflow/blob/cce45c208630830b28a32642bca345d3d75749a1/mlflow/tracking/client.py#L908) API.

### Motivation

Improve logging performance and reduce the number of API calls.

### Pitch

`MLFlowLogger` currently executes individual API calls for every hyperparameter (in [`log_hyperparameters`](https://github.com/Lightning-AI/lightning/blob/a86584d6dd4d50388c7dcef4f3854b0e8355b346/src/pytorch_lightning/loggers/mlflow.py#L240)) and metric (in [`log_metrics`](https://github.com/Lightning-AI/lightning/blob/a86584d6dd4d50388c7dcef4f3854b0e8355b346/src/pytorch_lightning/loggers/mlflow.py#L253)) using `for` loops. By refactoring these two methods to use `log_batch`, we could reduce the number and overhead of MLFlowLogger API calls.

I'd be happy to open a PR for this.

cc @borda

### PR DESCRIPTION
**Pull request by schmidt-jake**
## What does this PR do?

Fixes #15900.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)~
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?~
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)~

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified~ (I don't seem to have permissions for this)

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by schmidt-jake** [comment]
To avoid import errors, I hid some `mlflow.entities` imports inside the test bodies. If someone could shine some light on how to better structure (or mock) things, I will be happy to make changes!

File: src/pytorch_lightning/loggers/mlflow.py [hunk 1]
```diff
@@ -240,20 +241,26 @@ def experiment_id(self) -> Optional[str]:
     def log_hyperparams(self, params: Union[Dict[str, Any], Namespace]) -> None:
         params = _convert_params(params)
         params = _flatten_dict(params)
+        # params = [Param(k, v) for k, v in params.items()]
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:245]: is this a debugging statement we want to remove?

File: src/pytorch_lightning/loggers/mlflow.py [hunk 2]
```diff
@@ -240,20 +241,26 @@ def experiment_id(self) -> Optional[str]:
...
+        params_list: List[Param] = []
+
         for k, v in params.items():
+            # FIXME: mlflow 1.28 allows up to 500 characters: https://github.com/mlflow/mlflow/releases/tag/v1.28.0
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:249]: Are you planning to address this in this PR? I suggest to do it separately.

File: src/pytorch_lightning/loggers/mlflow.py [hunk 3]
```diff
@@ -269,8 +276,9 @@ def log_metrics(self, metrics: Mapping[str, float], step: Optional[int] = None)
                     category=RuntimeWarning,
                 )
                 k = new_k
+            metrics_list.append(Metric(key=k, value=v, timestamp=timestamp_ms, step=step or 0))
 
-            self.experiment.log_metric(self.run_id, k, v, timestamp_ms, step)
+        self.experiment.log_batch(run_id=self.run_id, metrics=metrics_list)
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:282]: Would you mind verifying in which version of mlflow this log_batch API was introduced?

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
> To avoid import errors, I hid some mlflow.entities imports inside the test bodies. If someone could shine some light on how to better structure (or mock) things, I will be happy to make changes!

Your solution would work assuming the package mlflow is installed. However, the tests there should run with mocked mlflow objects without the package being installed. I pushed a commit to demonstrate how to do it.

File: src/pytorch_lightning/loggers/mlflow.py (hunk 2 above)
Comment (schmidt-jake) [reply src/pytorch_lightning/loggers/mlflow.py:248]: I wasn't planning to address this in this PR, but figured adding a FIXME comment would still be appreciated for future devs — or should I remove it?

I'd be happy to address this in its own PR, though I'd have to study the patterns in place for managing logic conditional on version introspection.

Comment (awaelchli) [reply src/pytorch_lightning/loggers/mlflow.py:248]:
```suggestion
            # TODO: mlflow 1.28 allows up to 500 characters: https://github.com/mlflow/mlflow/releases/tag/v1.28.0
```

File: src/pytorch_lightning/loggers/mlflow.py (hunk 3 above)
Comment (awaelchli) [reply src/pytorch_lightning/loggers/mlflow.py:281]: > I was today years old when I learned I can query for text in github release notes.

Yes I could have done this myself :(( I'm sorry I bothered you for this.

In this case I agree we should set the minimum version to 1.0.0 (it is about as old as Lightning). And then we should be good :)

File: requirements/pytorch/loggers.info [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # all supported loggers. this list is here as a reference, but they are not installed in CI
 neptune-client
 comet-ml
-mlflow
+mlflow>=1
```
Comment (awaelchli) [inline requirements/pytorch/loggers.info:4]: @Borda I couldn't find where we are specifying the minimum version for loggers. Let's add it here?

Comment (awaelchli) [inline requirements/pytorch/loggers.info:4]:
```suggestion
mlflow>=1.0.0
```

Comment (carmocca) [reply requirements/pytorch/loggers.info:4]: Remember that we dont install these dependencies in CI. If our logger implentation does not work in lt 1.0, then it should be checked in the code at init time. For example by using RequirementCache

Comment (schmidt-jake) [reply requirements/pytorch/loggers.info:4]: Ok, I took a stab at this: https://github.com/Lightning-AI/lightning/pull/15915/commits/3428e5dbd6b2554dd2506b0d67c54f94569f5aa8

Let me know what you think!

File: src/pytorch_lightning/loggers/mlflow.py (hunk 3 above)
Comment (schmidt-jake) [reply src/pytorch_lightning/loggers/mlflow.py:285]: No, I'm glad you asked! I should have done it originally.

I'm searching around and the only reference to the mlflow dependency is in `loggers.info`, which I updated.

**Comment by schmidt-jake** [comment]
@Borda @carmocca looks like tests are all passing now! Could you please review?

File: src/pytorch_lightning/loggers/mlflow.py [hunk 5]
```diff
@@ -151,6 +154,9 @@ def __init__(
             raise ModuleNotFoundError(
                 "You want to use `mlflow` logger which is not installed yet, install it with `pip install mlflow`."
             )
+        if not RequirementCache("mlflow>=1.0.0"):
+            # we require the log_batch APIs that were introduced in mlflow 1.0.0
+            raise RuntimeError("Incompatible mlflow version")
```
Comment (awaelchli) [inline src/pytorch_lightning/loggers/mlflow.py:159]:
```suggestion
            raise RuntimeError("`MLFlowLogger` requires mlflow >= 1.0.0. Hint: Run `pip install -U mlflow`")
```

**Review by awaelchli — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
